File: Java-Kotlin/src/main/java/nowcoder/nc136.java
package nowcoder;

import common.datastructure.TreeNode;

import java.util.*;

public class nc136 {

    public int[] solve(int[] pre, int[] in) {
        // write code here
        if (pre == null || pre.length == 0 || in == null || in.length == 0 || pre.length != in.length) {
            return new int[]{};
        }

        TreeNode root = buildTree(pre, in);
        return getRightView(root);
    }

    private int[] getRightView(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        List<Integer> list = new ArrayList<>();

        TreeNode tmp = null;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                tmp = queue.poll();
                if (tmp.left != null) {
                    queue.offer(tmp.left);
                }
                if (tmp.right != null) {
                    queue.offer(tmp.right);
                }
            }
            list.add(tmp.val);
        }
        int[] res = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            res[i] = list.get(i);
        }
        return res;
    }

    private TreeNode buildTree(int[] pre, int[] in) {
        if (pre.length == 0 || in.length == 0) {
            return null;
        }

        TreeNode root = new TreeNode(pre[0]);
        for (int i = 0; i < in.length; i++) {
            if (in[i] == pre[0]) {
                root.left = buildTree(Arrays.copyOfRange(pre, 1, i + 1), Arrays.copyOfRange(in, 0, i));
                root.right = buildTree(Arrays.copyOfRange(pre, i + 1, pre.length), Arrays.copyOfRange(in, i + 1, in.length));
                break;
            }
        }
        return root;
    }
}

File: Go-Solutions/nowcoder/nc136.go
package nowcoder

func solve136(pre []int, in []int) []int {
	// write code here
	if len(pre) != len(in) || len(in) == 0 || len(pre) == 0 {
		return []int{}
	}

	root := buildTree136(pre, in)
	return getRightView136(root)
}

func buildTree136(pre []int, in []int) *TreeNode {
	if len(pre) == 0 || len(in) == 0 {
		return nil
	}

	root := &TreeNode{pre[0], nil, nil}
	for i := 0; i < len(in); i++ {
		if in[i] == pre[0] {
			root.Left = buildTree136(pre[1:i+1], in[0:i])
			root.Right = buildTree136(pre[i+1:], in[i+1:])
			break
		}
	}
	return root
}

func getRightView136(root *TreeNode) []int {
	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	res := make([]int, 0)
	var tmp *TreeNode = nil
	for len(queue) != 0 {
		size := len(queue)
		for i := 0; i < size; i++ {
			tmp = queue[0]
			queue = queue[1:]
			if tmp.Left != nil {
				queue = append(queue, tmp.Left)
			}
			if tmp.Right != nil {
				queue = append(queue, tmp.Right)
			}
		}
		res = append(res, tmp.Val)
	}
	return res
}
